_posts/2021-12-3-arrow-duckdb.md

```diff
@@ -0,0 +1,383 @@
+```r
+library(duckdb)
+library(arrow)
+library(dplyr)
+
+# Reads Parquet File to an Arrow Table
+arrow_table <- arrow::read_parquet("integers.parquet", as_data_frame = FALSE)
+
+# Gets Database Connection
+con <- dbConnect(duckdb::duckdb())
+
+# Registers arrow table as a DuckDB view
+arrow::to_duckdb(arrow_table, table_name = "arrow_table", con = con)
+
+# we can run a SQL query on this and print the result
+print(dbGetQuery(con, "SELECT SUM(data) FROM arrow_table WHERE data > 50"))
+
+# Transforms Query Result from DuckDB to Arrow Table
+result <- dbSendQuery(con, "SELECT * FROM arrow_table")
+```
+
+### Streaming Data from/to Arrow
+In the previous section, we depicted how to interact with Arrow tables. However, Arrow also allows users to interact with the data in a streaming fashion. Either consuming it (e.g., from an Arrow Dataset) or producing it (e.g., returning a RecordBatchReader). And of course, DuckDB is able to consume Datasets and produce RecordBatchReaders. This example uses the NYC Taxi Dataset, stored in Parquet files partitioned by year and month, which we can download through the Arrow R package:
+```R
+arrow::copy_files("s3://ursa-labs-taxi-data", "nyc-taxi")
+```
+
+
+#### Python
+```py
+# Reads dataset partitioning it in year/month folder
+nyc_dataset = ds.dataset('nyc-taxi/', partitioning=["year", "month"])
+
+# Gets Database Connection
+con = duckdb.connect()
+
+query = con.execute("SELECT * FROM nyc")
+# DuckDB's queries can now produce a Record Batch Reader
+record_batch_reader = query.fetch_record_batch()
+# Which means we can stream the whole query per batch.
+# This retrieves the first batch
+chunk = record_batch_reader.read_next_batch()
+```
+#### R
+```r
+# Reads dataset partitioning it in year/month folder
+nyc_dataset = open_dataset("nyc-taxi/", partitioning = c("year", "month"))
+
+# Gets Database Connection
+con <- dbConnect(duckdb::duckdb())
+
+# We can use the same function as before to register our arrow dataset
+duckdb::duckdb_register_arrow(con, "nyc", nyc_dataset)
+
+res <- dbSendQuery(con, "SELECT * FROM nyc", arrow = TRUE)
+# DuckDB's queries can now produce a Record Batch Reader
+record_batch_reader <- duckdb::duckdb_fetch_record_batch(res)
+
+# Which means we can stream the whole query per batch.
+# This retrieves the first batch
+cur_batch <- record_batch_reader$read_next_batch()
+```
+
+The preceding R code shows in low-level detail how the data is streaming. We provide the helper `to_arrow()` in the Arrow package which is a wrapper around this that makes it easy to incorporate this streaming into a dplyr pipeline. [^1]
+
+## Benchmark Comparison
+
+Here we demonstrate in a simple benchmark the performance difference between querying Arrow datasets with DuckDB and querying Arrow datasets with Pandas.
+For both the Projection and Filter pushdown comparison, we will use Arrow tables. That is due to Pandas not being capable of consuming Arrow stream objects.
```

Comment:
@pdet could we add information about the specs of / what kind of machine these were run on, as well as if you DuckDB was taking advantage of parallelism while you ran it (IIRC, the parallelism defaults to off with DuckDB)?

h/t @jorisvandenbossche

Reply:
parallelism is now on by default, and machine I think recent m1 MacBook Pro

Reply:
For the machine I used the diamonds for the nyc one
https://www.monetdb.org/wiki/Scilens-configuration-standard
And m1 for the tpch